File: 346_moving_average_from_data_stream.java
class MovingAverage {

    Deque<Integer> queue;
    int size;

    public MovingAverage(int size) {
        this.queue = new LinkedList<>();
        this.size = size;
    }

    public double next(int val) {
        if (queue.size() == this.size) {
            queue.pollFirst();
        }
        queue.offerLast(val);
        double result = 0.0;
        for (int num: queue) result += (double)num;
        return result / queue.size();
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */

File: python/346_moving_average_from_data_stream.py
class MovingAverage:

    def __init__(self, size: int):
        """
        Initialize your data structure here.
        """
        self.sum = 0
        self.size = size
        self.q = []

    def next(self, val: int) -> float:
        if len(self.q) == self.size:
            self.sum -= self.q.pop(0)
        self.sum += val
        self.q.append(val)
        return self.sum / len(self.q)

# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)
